File: app/src/main/java/com/example/purepawapp/ui/order/OrderHistoryAdapter.java
package com.example.purepawapp.ui.order;

import android.graphics.Typeface;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.Order;
import com.example.purepawapp.databinding.ItemOrderHistoryBinding;
import com.example.purepawapp.ui.admin.AdminStatusUi;
import com.example.purepawapp.ui.admin.StatusStyle;
import com.example.purepawapp.util.CurrencyUtils;

import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.function.Consumer;

public class OrderHistoryAdapter extends ListAdapter<Order, OrderHistoryAdapter.ViewHolder> {

    private final Consumer<Order> onClick;
    private final SimpleDateFormat dateFormat = new SimpleDateFormat("dd 'Tháng' MM, yyyy", new Locale("vi", "VN"));

    private static final DiffUtil.ItemCallback<Order> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull Order oldItem, @NonNull Order newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Order oldItem, @NonNull Order newItem) {
            return oldItem.getStatus().equals(newItem.getStatus()) && oldItem.getTotalAmount() == newItem.getTotalAmount();
        }
    };

    public OrderHistoryAdapter(Consumer<Order> onClick) {
        super(DIFF_CALLBACK);
        this.onClick = onClick;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemOrderHistoryBinding binding = ItemOrderHistoryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemOrderHistoryBinding binding;

        ViewHolder(ItemOrderHistoryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(Order order) {
            String orderCode = order.getOrderCode();
            binding.tvOrderId.setText(orderCode != null && !orderCode.isBlank() ? orderCode : "#" + order.getId().substring(0, Math.min(8, order.getId().length())));
            binding.tvDate.setText(dateFormat.format(order.getCreatedAt()));

            StatusStyle style = AdminStatusUi.orderStatusStyle(order.getStatus());
            binding.tvStatus.setText(style.getLabel());
            binding.tvStatus.setTextColor(binding.getRoot().getResources().getColor(style.getTextColorRes(), null));
            binding.tvStatus.setBackgroundTintList(binding.getRoot().getResources().getColorStateList(style.getBgColorRes(), null));

            int itemCount = 0;
            for (var item : order.getItems()) itemCount += item.getQuantity();
            String prefix = itemCount + " sản phẩm • ";
            String full = prefix + CurrencyUtils.toVndString(order.getTotalAmount());
            SpannableString spannable = new SpannableString(full);
            spannable.setSpan(new ForegroundColorSpan(binding.getRoot().getResources().getColor(R.color.pp_primary, null)),
                    prefix.length(), full.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            spannable.setSpan(new StyleSpan(Typeface.BOLD), prefix.length(), full.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            binding.tvSummary.setText(spannable);
            binding.tvSummary.setTextColor(binding.getRoot().getResources().getColor(R.color.pp_text_primary, null));

            binding.getRoot().setOnClickListener(v -> onClick.accept(order));
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/order/OrderHistoryAdapter.kt
package com.example.purepawapp.ui.order

import android.text.SpannableString
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.purepawapp.R
import com.example.purepawapp.data.model.Order
import com.example.purepawapp.databinding.ItemOrderHistoryBinding
import com.example.purepawapp.ui.admin.orderStatusStyle
import com.example.purepawapp.util.toVndString
import java.text.SimpleDateFormat
import java.util.Locale

class OrderHistoryAdapter(
    private val onClick: (Order) -> Unit
) : ListAdapter<Order, OrderHistoryAdapter.ViewHolder>(DIFF_CALLBACK) {

    private val dateFormat = SimpleDateFormat("dd 'Tháng' MM, yyyy", Locale("vi", "VN"))

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemOrderHistoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class ViewHolder(private val binding: ItemOrderHistoryBinding) : RecyclerView.ViewHolder(binding.root) {
        fun bind(order: Order) {
            binding.tvOrderId.text = order.orderCode.ifBlank { "#${order.id.take(8)}" }
            binding.tvDate.text = dateFormat.format(order.createdAt)

            val style = orderStatusStyle(order.status)
            binding.tvStatus.text = style.label
            binding.tvStatus.setTextColor(binding.root.resources.getColor(style.textColorRes, null))
            binding.tvStatus.backgroundTintList = binding.root.resources.getColorStateList(style.bgColorRes, null)

            val itemCount = order.items.sumOf { it.quantity }
            val prefix = "$itemCount sản phẩm • "
            val full = prefix + order.totalAmount.toVndString()
            val spannable = SpannableString(full)
            spannable.setSpan(
                ForegroundColorSpan(binding.root.resources.getColor(R.color.pp_primary, null)),
                prefix.length,
                full.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spannable.setSpan(
                StyleSpan(android.graphics.Typeface.BOLD),
                prefix.length,
                full.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            binding.tvSummary.text = spannable
            binding.tvSummary.setTextColor(binding.root.resources.getColor(R.color.pp_text_primary, null))

            binding.root.setOnClickListener { onClick(order) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<Order>() {
            override fun areItemsTheSame(oldItem: Order, newItem: Order): Boolean = oldItem.id == newItem.id
            override fun areContentsTheSame(oldItem: Order, newItem: Order): Boolean = oldItem == newItem
        }
    }
}
